pkg/scheduler/controller/backupbucket/scheduler_reconciler.go

```diff
@@ -0,0 +1,229 @@
+		// 	return err
+		// }
+		// if verifySeedAvailability(seed) {
+		// 	schedulerLogger.Infof("Seed %s is available, ignoring further reconciliation.", *backupBucket.Spec.Seed)
+		// 	return nil
+		// }
+		// schedulerLogger.Infof("Seed %s is not available, we will schedule it on another seed", *backupBucket.Spec.Seed)
+	}
+	// If no Seed is referenced, we try to determine an adequate one.
+	seed, err := r.determineSeed(backupBucket)
+	if err != nil {
+		r.reportFailedScheduling(backupBucket, err)
+		return err
+	}
+
+	if err := r.updateBackupBucketToBeScheduledOntoSeed(backupBucket, seed.Name); err != nil {
+		if _, ok := err.(*common.AlreadyScheduledError); ok {
+			return nil
+		}
+		r.reportFailedScheduling(backupBucket, err)
+		return err
+	}
+
+	schedulerLogger.Infof("BackupBucket '%s' (Cloud Profile '%s', Region '%s') successfully scheduled to seed '%s' ", backupBucket.Name, backupBucket.Spec.Provider, backupBucket.Spec.Region, seed.Name)
+	r.reportSuccessfulScheduling(backupBucket, seed.Name)
+	return nil
+}
+
+// determineSeed finds the appropriate seed for backupBucket.
+// It finds the seed by filtering out list as per policy mentioned below:
+// 1. Filter out seeds marked for deletion
+// 2. Filter out seeds which are not available currently.
+// 3. Select a seed if both, it's cloud provider and region matches with backupBucket.
+// 4. If failed find seed in step 3, then select a seed with matching cloud provider.
+// 5. If still not found then, select any of remaining seed.
+// 6. Return error if none of the above step found seed.
+func (r *reconciler) determineSeed(backupBucket *gardencorev1alpha1.BackupBucket) (*gardenv1beta1.Seed, error) {
+	var (
+		candidatesWithMathingProfile    = make([]*gardenv1beta1.Seed, 0)
+		candidatesWithoutMathingProfile = make([]*gardenv1beta1.Seed, 0)
+		cloudProfileCache               = make(map[string]gardenv1beta1.CloudProvider) //cloudProfile name verses cloudProvider map
+	)
+
+	seeds := &gardenv1beta1.SeedList{}
+	if err := r.client.List(r.ctx, seeds); err != nil {
+		return nil, err
+	}
+
+	if len(seeds.Items) == 0 {
+		return nil, fmt.Errorf("no seed found for scheduling")
+	}
+	for _, seed := range seeds.Items {
+		if seed.DeletionTimestamp != nil || verifySeedAvailability(&seed) {
+			continue
+		}
+
+		provider, ok := cloudProfileCache[seed.Spec.Cloud.Profile]
+		if !ok {
+			cloudProfile := &gardenv1beta1.CloudProfile{}
+			if err := r.client.Get(r.ctx, kutil.Key(seed.Spec.Cloud.Profile), cloudProfile); err != nil {
+				return nil, err
+			}
+
+			provider, err := gardenhelper.DetermineCloudProviderInProfile(cloudProfile.Spec)
+			if err != nil {
+				return nil, err
+			}
+			cloudProfileCache[cloudProfile.Name] = provider
+		}
+		if backupBucket.Spec.Provider == string(provider) {
+			if seed.Spec.Cloud.Region == backupBucket.Spec.Region {
+				return &seed, nil
+			}
+			candidatesWithMathingProfile = append(candidatesWithMathingProfile, &seed)
```

Comment:
Same question as above: After GEP4, how will this look like?

Reply:
As per GEP-4 or  the post discussion about location for backupConfig, i'll just grab the backup details from seed. - Choose the seed which from which backupBucket configuration is take. i.e. choose seed from which backupBucket.spec.provider.type == seed.spec.backup.type and same for the region
- If that seed is unavailble, choose seed with only matching type. 
- if still not valid seed found, choose random seed  or with matching region (min-distance-strategy)

Reply:
> Choose the seed which from which backupBucket configuration is take. i.e. choose seed from which backupBucket.spec.provider.type == seed.spec.backup.type and same for the region

What if there are multiple results, i.e., multiple seed clusters of the same provider in the same region?

Reply:
> What if there are multiple results, i.e., multiple seed clusters of the same provider in the same region?

How about picking a random seed from the list of seed for the same provider and region?
Similar to
> if still not valid seed found, choose random seed or with matching region (min-distance-strategy)